lib/recursive-methods.rb

```diff
@@ -1,49 +1,105 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the value of n
+# Space complexity: O(n) where n is the value of n
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  elsif n <= 1
+    return 1
+  else 
+    return n * factorial(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of s
+# Space complexity: O(n) where n is the length of s
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length <= 1
+    return s 
+  else 
+    return s[s.length - 1] + reverse(s[0..-2])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of s
+# Space complexity: O(n) where n is the length of s
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length <= 1
+    return s 
+  else 
+    return s[s.length - 1] + reverse(s[0..-2])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is value of n
+# Space complexity: O(n) where n is value of n
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 1
+    return 0
+  else 
+    return 2 + bunny(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of s
+# Space complexity: O(n) where n is the length of s
```

Comment:
Because you are making new arrays with each iteration, these are O(n<sup>2</sup>)